### PR TITLE
Pytorch module

### PR DESCRIPTION
**Pull request by kohr-h**
This adds missing functionality for proper pytorch integration. The existing `autograd.Function` wrapper cannot be used directly since `pytorch` doesn't quite know how to deal with extra axes when the Function doesn't handle that, and a naive implementation results in gradients not being propagated beyond that layer.
This PR fixes that and adds examples and tests for the new feature.

Some things are also renamed and moved around.

### CODE REVIEW

File: odl/contrib/pytorch/operator.py [hunk 1]
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
...
+        for i in range(x_flat_xtra.data.shape[0]):
+            results.append(self.op_func(x_flat_xtra[i]))
+
+        # Reshape the resulting stack to the expected output shape
+        stack_flat_xtra = torch.stack(results)
+        return stack_flat_xtra.view(extra_shape + op_out_shape)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        op_name = self.op_func.operator.__class__.__name__
```
Comment (adler-j) [inline odl/contrib/pytorch/operator.py]: Why not repr the actual operator?

File: odl/contrib/pytorch/test/test_operator.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+dtype = simple_fixture('dtype', ['float32', 'float64'])
+use_cuda_params = [False]
+if torch.cuda.is_available():
+    use_cuda_params.append(True)
+use_cuda = simple_fixture('use_cuda', use_cuda_params)
+shape = simple_fixture('shape', [(3,), (2, 3), (2, 2, 3)])
+
+
+def test_autograd_function_forward(dtype, use_cuda):
```
Comment (adler-j) [inline odl/contrib/pytorch/test/test_operator.py]: Very nice tests!

File: odl/contrib/pytorch/test/test_operator.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    x = torch.from_numpy(np.ones(3, dtype=dtype))
+    if use_cuda:
+        x = x.cuda()
+    x_var = autograd.Variable(x)
+
+    # Evaluate torch operator
+    res_var = torch_op(x_var)
+
+    # ODL result
+    odl_res = odl_op(x.cpu().numpy())
```
Comment (adler-j) [inline odl/contrib/pytorch/test/test_operator.py]: We should verify that if `use_cuda` is true, the result is still on CUDA

**Comment by adler-j** [comment]
By the way, have you managed to re-implement the learned primal dual article using pytorch?

File: odl/contrib/pytorch/operator.py [hunk 4]
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
...
+
+    def forward(self, x):
+        """Compute forward-pass of this module on ``x``.
+
+        Parameters
+        ----------
+        x : `torch.autograd.variable.Variable`
+            Input of this layer. The contained tensor must have shape
+            ``extra_shape + operator.domain.shape``, and
+            ``len(extra_shape)`` must be at least 1 (batch axis).
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/pytorch/operator.py]: That would just be an extra axis. This module is a bit like a linear layer that only understands input shape, output shape and batch axis. But in contrast to the linear layer, which can literally only act with one extra axis, this one supports any number of extra axes (batch, channel, you name it), simply by flattening them and then looping. Since there are no parameters to learn, any extra axis looks the same.

[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/pytorch/operator.py]: This made 100% sense, try to get some of that into the doc!

File: odl/contrib/pytorch/examples/operator_autograd.py [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
+``torch.autograd.Function`` that supports forward evaluation on
+pytorch ``Tensor`` objects and backpropagation using automatic
+differentiation.
+"""
 
 import numpy as np
 import torch
 import odl
-from odl.contrib.pytorch import TorchOperator
+from odl.contrib import pytorch
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/pytorch/examples/operator_autograd.py]: Okay, I'll go for something like `odl_torch` then. Had myself a bit doubts about `pytorch` :-)

File: odl/contrib/pytorch/examples/operator_module.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+"""Demonstration of ODL operators as pytorch modules.
+
+This example shows how to wrap an ODL ``Operator`` as a
+``torch.nn.Module`` that can be used as a layer in a Neural Network.
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/pytorch/examples/operator_module.py]: You mean the contrast to `autograd.Function`?

File: odl/contrib/pytorch/operator.py [hunk 7]
```diff
@@ -63,40 +73,52 @@ def forward(self, input):
...
-        [torch.DoubleTensor of size 1]
+        [torch.FloatTensor of size 1]
         """
+        # TODO: batched evaluation
         if not self.operator.is_linear:
             # Only needed for nonlinear operators
             self.save_for_backward(input)
 
-        input_arr = input.numpy()
+        # TODO: use GPU memory directly if possible
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/pytorch/operator.py]: Yeah well, for that the latter needs to exist :-)

File: odl/contrib/pytorch/operator.py [hunk 8]
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
...
+            The ODL operator to be wrapped. For gradient computations to
+            work, ``operator.derivative(x).adjoint`` must be implemented.
+
+        Examples
+        --------
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype='float32')
+        >>> odl_op = odl.MatrixOperator(matrix)
+        >>> torch_op = OperatorAsModule(odl_op)
+        >>> torch_op.op_func.operator is odl_op
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/pytorch/operator.py]: OK

File: odl/contrib/pytorch/operator.py [hunk 9]
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
...
+        >>> op_mod(x)
+        Variable containing:
+        (0 ,.,.) =
+           0   0
+           1   2
+        <BLANKLINE>
+        (1 ,.,.) =
+           2   4
+           3   6
+        <BLANKLINE>
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/pytorch/operator.py]: Dunno, I do want to have a non-square batch shape, and I think 3x2x2 is the smallest I can get. The `<BLANKLINE>` things are not displayed in the rendered docs btw. They're just for doctest.

File: odl/contrib/pytorch/operator.py (hunk 1 above)
Comment (kohr-h) [reply odl/contrib/pytorch/operator.py]: Because of the self-printing of composed modules. For example:
```
NNFBP (
  (conv_det): Conv2d(1, 32, kernel_size=(1, 11), stride=(1, 1), padding=(0, 10))
  (backproj): OperatorAsModule(RayBackProjection) ((50, 256) -> (128, 128))
  (conv1_vol): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(2, 2))
  (conv2_vol): Conv2d(32, 1, kernel_size=(3, 3), stride=(1, 1), padding=(2, 2))
)
```
But I can check if that works with `__str__`.

**Comment by adler-j** [comment]
Really minor stuff left to get this in, lets get it done.

**Comment by kohr-h** [comment]
Yes, I'll look into this tomorrow when I'm back in office. I don't feel like compiling pytorch again on my laptop :-)

File: odl/contrib/torch/examples/operator_module.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+# Wrap ODL operator as `Module`
+op_layer = pytorch.operator.OperatorAsModule(odl_op)
+
+# Test with some inputs. We need to add at least one fake batch axis
+inp = autograd.Variable(torch.ones(3))[None, ...]
+print('Operator layer evaluated on a 1x3 tensor variable:')
+print(op_layer(inp))
+
+inp = autograd.Variable(torch.ones(3))[None, None, ...]
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/torch/examples/operator_module.py]: I don't know if examples are the place to do this. Extra batch and channel axes is all that people likely need to use, so I think I'll keep it at this.

**Comment by kohr-h** [comment]
Rebased and fixed the stuff. There was something I wanted to fix as well but don't remember now. But I'm gonna merge this now and fix the other things in the next PR.
